lib/index.js

```diff
@@ -10,16 +10,30 @@ module.exports = function(schema, options, version) {
   // clone main level of options (does not clone deeper)
   options = utils.highClone(options);
 
+  let generateType;
+  if (options.withDiscriminators && typeof options.type === 'function') {
+    generateType = options.type;
+  }
+
   /**
    * Retrieve model options to ElasticSearch
    * static function
    * returns {Object}
    */
+
   function esOptions() {
     if (!options.index) {
       options.index = this.collection.name;
     }
-    if (!options.type) {
+
+    const kind = this.constructor.modelName || this.modelName;
```

Comment:
kind should be defined inside if

Reply:
May be generateType also should be inside your if?

Reply:
I think, `generateType` should be outside, because it came from base schema, so if we'll put it inside `esOptions` for instance, then child schemas won't be able to generate the type.